transformations/contraction_expansions/transformation.py

```diff
@@ -0,0 +1,171 @@
+            "where's": "where is",
+            "where've": "where have",
+            "who'll": "who will",
+            "who's": "who is",
+            "who've": "who have",
+            "why's": "why is",
+            "won't": "will not",
+            "would've": "would have",
+            "wouldn't": "would not",
+            "you'd": "you would",
+            "you'd've": "you would have",
+            "you'll": "you will",
+            "you're": "you are",
+            "you've": "you have",
+        }
+        self.reverse_contraction_map = {value:key for key, value in self.contraction_map.items()}
+    
+
+    def contract(self, sentence):
+        """Contract expanded contractions in a sentence (if any)
+        Parameters
+        ----------
+        sentence : str
+            input string
+        Returns
+        -------
+        string
+            String with contractions contracted (if any)
+        """
+        
+        reverse_contraction_pattern = re.compile(
+            r"\b({})\b ".format("|".join(self.reverse_contraction_map.keys())),
+            flags=re.IGNORECASE | re.DOTALL,
+        )
+
+        def cont(possible):
+            match = possible.group(1)
+            first_char = match[0]
+            expanded_contraction = self.reverse_contraction_map.get(
+                match, self.reverse_contraction_map.get(match.lower())
+            )
+            expanded_contraction = first_char + expanded_contraction[1:] + " "
+            return expanded_contraction
+
+        return reverse_contraction_pattern.sub(cont, sentence)
+
+    def expand_contractions(self, sentence):
+        """Expands contractions in a sentence (if any)
+        Parameters
+        ----------
+        sentence : str
+            input string
+        Returns
+        -------
+        string
+            String with contractions expanded (if any)
+        """
+        
+
+        contraction_pattern = re.compile(
+            r"\b({})\b".format("|".join(self.contraction_map.keys())),
+            flags=re.IGNORECASE | re.DOTALL,
+        )
+
+        def expand_match(contraction):
+            match = contraction.group(0)
+            first_char = match[0]
+            expanded_contraction = self.contraction_map.get(
+                match, self.contraction_map.get(match.lower())
+            )
+            expanded_contraction = first_char + expanded_contraction[1:]
+            return expanded_contraction
+
+        return contraction_pattern.sub(expand_match, sentence)
+
+    def contractions(self, sentence):
+        """Perturbation functions, contracts and expands contractions if present
+        Parameters
+        ----------
+        sentence : str
+            input
+        Returns
+        -------
+        list
+            List of strings with contractions expanded or contracted, or []
+        """
+        expanded = [self.expand_contractions(sentence), self.contract(sentence)]
+        return [t for t in expanded if t != sentence]
```

Comment:
One more question, if the input sentence contains both expanded and contracted segments, it would return both fully expanded and contracted sentences, right?

For example:
Input: "He is a doctor, but she's a teacher"
The output would be: ["He is a doctor, but she is a teacher", "He's a doctor, but she's a teacher"]

Reply:
Also, I think maybe it's fine to return the original sentence rather than an empty list when you **cannot** find any expanded or contracted phrases. Because one might want to apply this transformation to the dataset and the current implementation might result in empty inputs in the transformed data.

I guess the best way is to remove the original sentence if there **is** a successful expansion or contraction. Otherwise, you can just return the unchanged input.

Reply:
> One more question, if the input sentence contains both expanded and contracted segments, it would return both fully expanded and contracted sentences, right?
> 
> For example:
> Input: "He is a doctor, but she's a teacher"
> The output would be: ["He is a doctor, but she is a teacher", "He's a doctor, but she's a teacher"]

Yes it will, have added another test case to clear things out

Reply:
> Also, I think maybe it's fine to return the original sentence rather than an empty list when you **cannot** find any expanded or contracted phrases. Because one might want to apply this transformation to the dataset and the current implementation might result in empty inputs in the transformed data.
> 
> I guess the best way is to remove the original sentence if there **is** a successful expansion or contraction. Otherwise, you can just return the unchanged input.

Thanks for pointing this out, I wasn't very sure about this case , will update the code